### ISSUE TITLE
Allow infix syntax in `data` declarations

### ISSUE THREAD

**Issue by lexi-lambda** [lexi-lambda/hackett#11]
These sorts of things should be valid:

```racket
(data {a :~: b} ...)
(data (List a)
  {a :: (List a)}
  nil)
```

**Comment by gelisam** [comment]
I got `{a :: (List a)}` working! Can you mark the issue as 'in-progress' so that others don't duplicate the work?

**Comment by lexi-lambda** [comment]
Fixed by 7add7bc4a89c73a70ccc277a8f15d3463dbe1da8.

**Comment by lexi-lambda** [comment]
Reintroduced by the new implementation, with a twist.

The new implementation allows specifying associativity. Precedence is not yet supported, and the syntax is all very tentative, but the basic idea is that you can define left-associative or right-associative operators by writing something like this:

```racket
(def :: #:fixity right
  cons)
```

And then when you write this:

```racket
{1 :: 2 :: 3 :: nil}
```

…it gets parsed with the correct precedence.

Unfortunately, writing `::` as a separate definition from `cons` means we can’t use `::` in `case`, so we really watch to attach the fixity annotations to the `data` declarations themselves. This creates a small dilemma, since it’s not immediately clear to me where in the `data` declarations fixity annotations should go.

Should they look like this?

```racket
(data {a :<|>: b} #:fixity right
  {a :<||>: b} #:fixity right)
```

I’m not sure how else to do it without adding another grouping construct in `data` declarations.
